File: AngularJS/src/org/angularjs/cli/AngularCLIProjectGenerator.java
package org.angularjs.cli;

import com.intellij.execution.RunManager;
import com.intellij.execution.RunnerAndConfigurationSettings;
import com.intellij.execution.configurations.ConfigurationType;
import com.intellij.execution.filters.Filter;
import com.intellij.ide.util.projectWizard.ModuleNameLocationSettings;
import com.intellij.ide.util.projectWizard.SettingsStep;
import com.intellij.javascript.debugger.execution.JavaScriptDebugConfiguration;
import com.intellij.javascript.debugger.execution.JavascriptDebugConfigurationType;
import com.intellij.javascript.karma.execution.KarmaConfigurationType;
import com.intellij.javascript.karma.execution.KarmaRunConfiguration;
import com.intellij.javascript.karma.execution.KarmaRunSettings;
import com.intellij.javascript.karma.util.KarmaUtil;
import com.intellij.javascript.nodejs.interpreter.NodeJsInterpreter;
import com.intellij.javascript.nodejs.interpreter.NodeJsInterpreterRef;
import com.intellij.javascript.nodejs.util.NodePackage;
import com.intellij.javascript.nodejs.util.NodePackageDescriptor;
import com.intellij.javascript.protractor.ProtractorConfigurationType;
import com.intellij.javascript.protractor.ProtractorRunConfiguration;
import com.intellij.javascript.protractor.ProtractorRunSettings;
import com.intellij.javascript.protractor.ProtractorUtil;
import com.intellij.lang.javascript.boilerplate.NpmPackageProjectGenerator;
import com.intellij.lang.javascript.boilerplate.NpxPackageDescriptor;
import com.intellij.lang.javascript.buildTools.npm.rc.NpmCommand;
import com.intellij.lang.javascript.buildTools.npm.rc.NpmConfigurationType;
import com.intellij.lang.javascript.buildTools.npm.rc.NpmRunConfiguration;
import com.intellij.lang.javascript.buildTools.npm.rc.NpmRunSettings;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.module.ModuleManager;
import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ContentEntry;
import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VfsUtilCore;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.TextAccessor;
import com.intellij.util.ArrayUtil;
import com.intellij.util.Consumer;
import com.intellij.util.PathUtil;
import com.intellij.xml.util.XmlStringUtil;
import icons.AngularJSIcons;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.io.File;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.regex.Pattern;

/**
 * @author Dennis.Ushakov
 */
public class AngularCLIProjectGenerator extends NpmPackageProjectGenerator {

  public static final String PACKAGE_NAME = "@angular/cli";
  private static final Pattern VALID_NG_APP_NAME = Pattern.compile("[a-zA-Z][0-9a-zA-Z]*(-[a-zA-Z][0-9a-zA-Z]*)*");

  @Nls
  @NotNull
  @Override
  public String getName() {
    return "Angular CLI";
  }

  @Override
  @NotNull
  public String getDescription() {
    return "The Angular CLI makes it easy to create an application that already works, right out of the box. It already follows our best practices!";
  }

  @Override
  @NotNull
  public Icon getIcon() {
    return AngularJSIcons.Angular2;
  }

  @Override
  protected void customizeModule(@NotNull VirtualFile baseDir, ContentEntry entry) {
    if (entry != null) {
      AngularJSProjectConfigurator.excludeDefault(baseDir, entry);
    }
  }

  @Override
  @NotNull
  protected String[] generatorArgs(@NotNull Project project, @NotNull VirtualFile baseDir) {
    return ArrayUtil.EMPTY_STRING_ARRAY;
  }

  @Override
  @NotNull
  protected String[] generatorArgs(@NotNull Project project, @NotNull VirtualFile baseDir, @NotNull Settings settings) {
    return new String[]{"new", baseDir.getName()};
  }

  @NotNull
  @Override
  protected Filter[] filters(@NotNull Project project, @NotNull VirtualFile baseDir) {
    return new Filter[]{new AngularCLIFilter(project, baseDir.getParent().getPath())};
  }

  @NotNull
  @Override
  protected String executable(@NotNull NodePackage pkg) {
    return ng(pkg.getSystemDependentPath());
  }

  @NotNull
  public static String ng(String path) {
    return path + File.separator + "bin" + File.separator + "ng";
  }

  @Override
  @NotNull
  protected String packageName() {
    return PACKAGE_NAME;
  }

  @Override
  @NotNull
  protected String presentablePackageName() {
    return "Angular &CLI:";
  }

  @NotNull
  @Override
  protected List<NpxPackageDescriptor.NpxCommand> getNpxCommands() {
    return Collections.singletonList(new NpxPackageDescriptor.NpxCommand(PACKAGE_NAME, "ng"));
  }

  @Override
  protected String validateProjectPath(@NotNull String path) {
    return Optional.ofNullable(validateFolderName(path, "Project"))
                   .orElseGet(() -> super.validateProjectPath(path));
  }

  @SuppressWarnings("deprecation")
  @NotNull
  @Override
  public GeneratorPeer<Settings> createPeer() {
    return new AngularCLIProjectGeneratorPeer();
  }

  @NotNull
  @Override
  protected File workingDir(Settings settings, @NotNull VirtualFile baseDir) {
    return VfsUtilCore.virtualToIoFile(baseDir).getParentFile();
  }


  @NotNull
  @Override
  protected Runnable postInstall(@NotNull Project project,
                                 @NotNull VirtualFile baseDir,
                                 File workingDir) {
    return () -> ApplicationManager.getApplication().executeOnPooledThread(() -> {
      super.postInstall(project, baseDir, workingDir).run();
      createRunConfigurations(project, baseDir);
    });
  }

  @Nullable
  private static String validateFolderName(String path, String label) {
    String fileName = PathUtil.getFileName(path);
    if (!VALID_NG_APP_NAME.matcher(fileName).matches()) {
      return XmlStringUtil.wrapInHtml(
        label + " name '" + fileName + "' is not valid. " + label + " name must " +
        "start with a letter, and must contain only alphanumeric characters or dashes. " +
        "When adding a dash the segment after the dash must also start with a letter."
      );
    }
    return null;
  }

  public static void createRunConfigurations(@NotNull Project project, @NotNull VirtualFile baseDir) {
    ApplicationManager.getApplication().executeOnPooledThread(
      () -> DumbService.getInstance(project).runReadActionInSmartMode(() -> {
        if (project.isDisposed()) {
          return;
        }
        RunManager runManager = RunManager.getInstance(project);

        String packageJsonPath = getPackageJson(baseDir);
        if (packageJsonPath == null) {
          return;
        }

        String nameSuffix = ModuleManager.getInstance(project).getModules().length > 1
                            ? " (" + baseDir.getName() + ")" : "";

        createJSDebugConfiguration(runManager, "Angular Application" + nameSuffix, "http://localhost:4200");
        createKarmaConfiguration(project, baseDir, runManager, "Tests" + nameSuffix);
        createProtractorConfiguration(project, baseDir, runManager, "E2E Tests" + nameSuffix);
        runManager.setSelectedConfiguration(
          createNpmConfiguration(packageJsonPath, runManager, "Angular CLI Server" + nameSuffix, "start"));
      }));
  }

  @Nullable
  private static String getPackageJson(@NotNull VirtualFile baseDir) {
    VirtualFile pkg = baseDir.findChild("package.json");
    if (pkg != null && !pkg.isDirectory()) {
      return pkg.getPath();
    }
    return null;
  }

  private static void createJSDebugConfiguration(@NotNull RunManager runManager, @NotNull String label, @NotNull String url) {
    createRunConfig(runManager, label, JavascriptDebugConfigurationType.getTypeInstance(),
                    (JavaScriptDebugConfiguration config) -> url.equals(config.getUri()),
                    (JavaScriptDebugConfiguration config) -> config.setUri(url));
  }

  @NotNull
  private static RunnerAndConfigurationSettings createNpmConfiguration(@NotNull String packageJsonPath,
                                                                       @NotNull RunManager runManager,
                                                                       @NotNull String label,
                                                                       @NotNull String scriptName) {
    NpmRunSettings runSettings = NpmRunSettings
      .builder()
      .setCommand(NpmCommand.RUN_SCRIPT)
      .setScriptNames(Collections.singletonList(scriptName))
      .setPackageJsonPath(packageJsonPath)
      .build();
    return createRunConfig(runManager, label, NpmConfigurationType.getInstance(),
                           (NpmRunConfiguration config) -> similar(config.getRunSettings(), runSettings),
                           (NpmRunConfiguration config) -> config.setRunSettings(runSettings));
  }

  private static void createKarmaConfiguration(@NotNull Project project,
                                               @NotNull VirtualFile baseDir,
                                               @NotNull RunManager runManager,
                                               @NotNull String label) {
    KarmaRunSettings runSettings = new KarmaRunSettings.Builder()
      .setKarmaPackage(findPackage(project, baseDir, KarmaUtil.ANGULAR_CLI__PACKAGE_NAME))
      .setConfigPath(findConfigFile(project, baseDir, KarmaUtil::listPossibleConfigFilesInProject))
      .setWorkingDirectory(baseDir.getPath())
      .build();
    createRunConfig(runManager, label, KarmaConfigurationType.getInstance(),
                    (KarmaRunConfiguration config) -> similar(config.getRunSettings(), runSettings),
                    (KarmaRunConfiguration config) -> config.setRunSettings(runSettings)
    );
  }

  private static void createProtractorConfiguration(@NotNull Project project,
                                                    @NotNull VirtualFile baseDir,
                                                    @NotNull RunManager runManager,
                                                    @NotNull String label) {
    ProtractorRunSettings runSettings = new ProtractorRunSettings.Builder()
      .setConfigFilePath(StringUtil.defaultIfEmpty(
        findConfigFile(project, baseDir, ProtractorUtil::listPossibleConfigFilesInProject), ""))
      .build();
    createRunConfig(runManager, label, ProtractorConfigurationType.getInstance(),
                    (ProtractorRunConfiguration config) -> similar(config.getRunSettings(), runSettings),
                    (ProtractorRunConfiguration config) -> config.setRunSettings(runSettings)
    );
    Optional.ofNullable(findPackage(project, baseDir, ProtractorUtil.PACKAGE_NAME))
            .ifPresent(pkg -> ProtractorUtil.setProtractorPackage(project, pkg));
  }

  private static boolean similar(@NotNull ProtractorRunSettings s1, @NotNull ProtractorRunSettings s2) {
    return s1.getConfigFileSystemDependentPath().equals(s2.getConfigFileSystemDependentPath());
  }

  private static boolean similar(@NotNull KarmaRunSettings s1, @NotNull KarmaRunSettings s2) {
    return s1.getConfigPathSystemDependent().equals(s2.getConfigPathSystemDependent())
           && Objects.equals(s1.getKarmaPackage(), s2.getKarmaPackage());
  }

  private static boolean similar(@NotNull NpmRunSettings s1, @NotNull NpmRunSettings s2) {
    return s1.getCommand().equals(s2.getCommand())
           && s1.getScriptNames().equals(s2.getScriptNames())
           && s1.getPackageJsonSystemDependentPath().equals(s2.getPackageJsonSystemDependentPath());
  }

  @SuppressWarnings("unchecked")
  @NotNull
  private static <T> RunnerAndConfigurationSettings createRunConfig(@NotNull RunManager runManager,
                                                                    @NotNull String label,
                                                                    @NotNull ConfigurationType configurationType,
                                                                    @NotNull Predicate<T> isSimilar,
                                                                    @NotNull Consumer<T> configurator) {
    return runManager
      .getConfigurationSettingsList(configurationType)
      .stream()
      .filter(settings -> isSimilar.test((T)settings.getConfiguration()))
      .findFirst()
      .orElseGet(() -> {
        RunnerAndConfigurationSettings settings =
          runManager.createConfiguration(label, configurationType.getConfigurationFactories()[0]);

        configurator.consume((T)settings.getConfiguration());

        runManager.addConfiguration(settings);
        return settings;
      });
  }

  @Nullable
  private static NodePackage findPackage(@NotNull Project project, @NotNull VirtualFile baseDir, @NotNull String packageName) {
    NodeJsInterpreter interpreter = NodeJsInterpreterRef.createProjectRef().resolve(project);
    NodePackageDescriptor descr = new NodePackageDescriptor(packageName);
    return descr.listAvailable(project, interpreter, baseDir, true)
                .stream()
                .filter(p -> p.getSystemIndependentPath().startsWith(baseDir.getPath())).findFirst().orElse(null);
  }

  @Nullable
  private static String findConfigFile(@NotNull Project project,
                                       @NotNull VirtualFile baseDir,
                                       Function<Project, List<VirtualFile>> listProvider) {
    return listProvider.apply(project)
                       .stream()
                       .filter(f -> f.getPath().startsWith(baseDir.getPath()))
                       .findFirst()
                       .map(f -> f.getPath())
                       .orElse(null);
  }

  private class AngularCLIProjectGeneratorPeer extends NpmPackageGeneratorPeer {

    private TextAccessor myContentRoot;

    @Override
    public void buildUI(@NotNull SettingsStep settingsStep) {
      super.buildUI(settingsStep);
      final ModuleNameLocationSettings field = settingsStep.getModuleNameLocationSettings();
      if (field != null) {
        myContentRoot = new TextAccessor() {
          @Override
          public void setText(@NotNull String text) {
            field.setModuleContentRoot(text);
          }

          @Override
          @NotNull
          public String getText() {
            return field.getModuleContentRoot();
          }
        };
      }
    }

    @Nullable
    @Override
    public ValidationInfo validate() {
      final ValidationInfo info = super.validate();
      if (info != null) {
        return info;
      }
      if (myContentRoot != null) {
        String message = validateFolderName(myContentRoot.getText(), "Content root folder");
        if (message != null) {
          return new ValidationInfo(message);
        }
      }
      return null;
    }
  }
}

File: Angular/src/org/angular2/cli/AngularCliProjectGenerator.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.cli

import com.intellij.execution.configurations.CommandLineTokenizer
import com.intellij.execution.filters.Filter
import com.intellij.ide.util.projectWizard.SettingsStep
import com.intellij.javascript.nodejs.util.NodePackage
import com.intellij.lang.javascript.boilerplate.NpmPackageProjectGenerator
import com.intellij.lang.javascript.boilerplate.NpxPackageDescriptor
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ReadAction
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.roots.ContentEntry
import com.intellij.openapi.ui.LabeledComponent
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.openapi.util.NlsContexts.DialogMessage
import com.intellij.openapi.util.Ref
import com.intellij.openapi.vfs.StandardFileSystems
import com.intellij.openapi.vfs.VfsUtilCore
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.platform.ProjectGeneratorPeer
import com.intellij.ui.TextAccessor
import com.intellij.util.ArrayUtilRt
import com.intellij.util.Consumer
import com.intellij.util.PathUtil
import com.intellij.util.text.SemVer
import com.intellij.util.ui.UIUtil
import com.intellij.xml.util.XmlStringUtil
import icons.AngularIcons
import org.angular2.lang.Angular2Bundle
import org.angular2.lang.Angular2LangUtil.ANGULAR_CLI_PACKAGE
import org.jetbrains.annotations.Nls
import java.awt.BorderLayout
import java.io.File
import java.util.regex.Pattern
import javax.swing.Icon
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JPanel

class AngularCliProjectGenerator : NpmPackageProjectGenerator() {

  override fun getId(): String {
    return "AngularCLI"
  }

  @Nls
  override fun getName(): String {
    return Angular2Bundle.message("angular.action.new-project.name")
  }

  override fun getDescription(): String {
    return Angular2Bundle.message("angular.action.new-project.description")
  }

  override fun getIcon(): Icon {
    return AngularIcons.Angular2
  }

  override fun customizeModule(baseDir: VirtualFile, entry: ContentEntry?) {
    entry?.addDefaultAngularExcludes(baseDir)
  }

  override fun generatorArgs(project: Project, baseDir: VirtualFile): Array<String> {
    return ArrayUtilRt.EMPTY_STRING_ARRAY
  }

  override fun generatorArgs(project: Project, baseDir: VirtualFile, settings: Settings): Array<String> {
    val ngSettings = settings as AngularCLIProjectSettings
    val result = ArrayList<String>()
    result.add("new")
    result.add(baseDir.name)
    val tokenizer = CommandLineTokenizer(ngSettings.options)
    while (tokenizer.hasMoreTokens()) {
      result.add(tokenizer.nextToken())
    }

    if (isPackageGreaterOrEqual(settings.myPackage, 7, 0, 0) && ngSettings.useDefaults) {
      if (result.none { param -> param == "--defaults" || param.startsWith("--defaults=") }) {
        result.add("--defaults")
      }
    }

    if (isPackageGreaterOrEqual(settings.myPackage, 16, 0, 0)) {
      if (result.none { param -> param == "--standalone" || param.startsWith("--standalone=") }) {
        result.add("--standalone=${ngSettings.useStandalone}")
      }
    }

    return ArrayUtilRt.toStringArray(result)
  }

  override fun filters(project: Project, baseDir: VirtualFile): Array<Filter> {
    return arrayOf(AngularCliFilter(project, baseDir.parent.path))
  }

  override fun executable(pkg: NodePackage): String {
    return ng(pkg.systemDependentPath)
  }

  override fun packageName(): String {
    return ANGULAR_CLI_PACKAGE
  }

  override fun presentablePackageName(): String {
    return Angular2Bundle.message("angular.action.new-project.presentable-package-name")
  }

  override fun getNpxCommands(): List<NpxPackageDescriptor.NpxCommand> {
    return listOf(NpxPackageDescriptor.NpxCommand(ANGULAR_CLI_PACKAGE, NG_EXECUTABLE))
  }

  override fun validateProjectPath(path: String): String? {
    return validateFolderName(path, Angular2Bundle.message("angular.action.new-project.label-project-name"))
           ?: super.validateProjectPath(path)
  }

  override fun createPeer(): ProjectGeneratorPeer<Settings> {
    return AngularCLIProjectGeneratorPeer()
  }

  override fun workingDir(settings: Settings, baseDir: VirtualFile): File {
    return VfsUtilCore.virtualToIoFile(baseDir).parentFile
  }


  override fun postInstall(
    project: Project,
    baseDir: VirtualFile,
    workingDir: File,
  ): Runnable {
    return Runnable {
      ApplicationManager.getApplication().executeOnPooledThread {
        super.postInstall(project, baseDir, workingDir).run()
        AngularCliUtil.createRunConfigurations(project, baseDir)
      }
    }
  }


  private inner class AngularCLIProjectGeneratorPeer : NpmPackageProjectGenerator.NpmPackageGeneratorPeer() {

    private var myContentRoot: TextAccessor? = null

    private lateinit var myOptionsTextField: SchematicOptionsTextField
    private lateinit var myUseDefaults: JCheckBox
    private lateinit var myUseStandalone: JCheckBox

    override fun createPanel(): JPanel {
      val panel = super.createPanel()

      myOptionsTextField = SchematicOptionsTextField(ProjectManager.getInstance().defaultProject,
                                                     emptyList(), UNKNOWN_VERSION)
      myOptionsTextField.setVariants(listOf(Option("test")))

      val component = LabeledComponent.create(
        myOptionsTextField, Angular2Bundle.message("angular.action.new-project.label-additional-parameters"))
      component.setAnchor(panel.getComponent(0) as JComponent)
      component.labelLocation = BorderLayout.WEST
      panel.add(component)

      myUseStandalone = JCheckBox(Angular2Bundle.message("angular.action.new-project.label-standalone"), true)
      panel.add(myUseStandalone)

      myUseDefaults = JCheckBox(Angular2Bundle.message("angular.action.new-project.label-defaults"), true)
      panel.add(myUseDefaults)


      return panel
    }

    override fun addExtraFields(settingsStep: SettingsStep) {
      val field = settingsStep.moduleNameLocationSettings
      if (field != null) {
        myContentRoot = object : TextAccessor {
          override fun setText(text: String) {
            field.moduleContentRoot = text
          }

          override fun getText(): String {
            return field.moduleContentRoot
          }
        }
      }
      settingsStep.addSettingsField(UIUtil.replaceMnemonicAmpersand(
        Angular2Bundle.message("angular.action.new-project.label-additional-parameters")), myOptionsTextField)
      settingsStep.addSettingsComponent(myUseStandalone)
      settingsStep.addSettingsComponent(myUseDefaults)
      packageField.addSelectionListener(Consumer { this.nodePackageChanged(it) })
      nodePackageChanged(packageField.selected)
    }

    override fun getSettings(): Settings {
      return AngularCLIProjectSettings(super.getSettings(), myUseDefaults.isSelected, myUseStandalone.isSelected, myOptionsTextField.text)
    }

    override fun validate(): ValidationInfo? {
      val info = super.validate()
      if (info != null) {
        return info
      }
      if (myContentRoot != null) {
        val message = validateFolderName(myContentRoot!!.text,
                                         Angular2Bundle.message("angular.action.new-project.label-content-root-folder"))
        if (message != null) {
          return ValidationInfo(message)
        }
      }
      return null
    }

    private fun nodePackageChanged(nodePackage: NodePackage) {
      ApplicationManager.getApplication().executeOnPooledThread {
        val options = Ref.create(emptyList<Option>())
        val cliVersion = Ref.create(UNKNOWN_VERSION)
        if (nodePackage.systemIndependentPath.endsWith("/node_modules/@angular/cli")) {
          var localFile = StandardFileSystems.local().findFileByPath(
            nodePackage.systemDependentPath)
          if (localFile != null) {
            localFile = localFile.parent.parent.parent
            try {
              val availableOptions =
                AngularCliSchematicsRegistryService.instance
                  .getSchematics(ProjectManager.getInstance().defaultProject, localFile!!, true, false)
                  .asSequence()
                  .filter { s -> "ng-new" == s.name }
                  .firstOrNull()
                  ?.let { schematic ->
                    val list = ArrayList(schematic.options)
                    list.add(createOption("verbose", "Boolean", false, "Adds more details to output logging."))
                    list.add(createOption("collection", "String", null, "Schematics collection to use"))
                    list.sortBy { it.name }
                    list
                  }
                ?: emptyList()

              options.set(availableOptions)
            }
            catch (e: Exception) {
              thisLogger().error("Failed to load schematics", e)
            }

            val packageVersion = nodePackage.version
            if (packageVersion != null) {
              cliVersion.set(packageVersion)
            }
          }
        }
        ReadAction.run<RuntimeException> { myOptionsTextField.setVariants(options.get(), cliVersion.get()) }
      }
    }

    private fun createOption(name: String, type: String, defaultVal: Any?, description: String): Option {
      val res = Option(name)
      res.type = type
      res.default = defaultVal
      res.description = description
      return res
    }
  }

  private class AngularCLIProjectSettings(
    settings: Settings,
    val useDefaults: Boolean,
    val useStandalone: Boolean,
    val options: String,
  )
    : Settings(settings.myInterpreterRef, settings.myPackage)

  companion object {
    const val NG_EXECUTABLE: String = "ng"
  }
}

private val NPX_PACKAGE_PATTERN = Pattern.compile("npx --package @angular/cli(?:@([0-9]+\\.[0-9]+\\.[0-9a-zA-Z-.]+))? ng")
private val VALID_NG_APP_NAME = Pattern.compile("[a-zA-Z][0-9a-zA-Z]*(-[a-zA-Z][0-9a-zA-Z]*)*")
private val UNKNOWN_VERSION = SemVer("0.0.0", 0, 0, 0)

private fun isPackageGreaterOrEqual(pkg: NodePackage, major: Int, minor: Int, patch: Int): Boolean {
  var ver: SemVer? = null
  if (pkg.name == ANGULAR_CLI_PACKAGE) {
    ver = pkg.version
  }
  else {
    val m = NPX_PACKAGE_PATTERN.matcher(pkg.systemIndependentPath)
    if (m.matches()) {
      ver = SemVer.parseFromText(m.group(1))
    }
  }
  return ver == null || ver.isGreaterOrEqualThan(major, minor, patch)
}

fun ng(path: String): String {
  return path + File.separator + "bin" + File.separator + AngularCliProjectGenerator.NG_EXECUTABLE
}

@DialogMessage
private fun validateFolderName(path: String, label: String): String? {
  val fileName = PathUtil.getFileName(path)
  return if (!VALID_NG_APP_NAME.matcher(fileName).matches()) {
    XmlStringUtil.wrapInHtml(
      Angular2Bundle.message("angular.action.new-project.wrong-folder-name", label, fileName)
    )
  }
  else null
}